### PR TITLE
chore: Introduce more aggressive caching

### PR DESCRIPTION
**Pull request by blt**
This commit introduces caching of `target/debug` into our jobs as well as the
use of `restore-keys` to hit cache even when `Cargo.lock` has changed. This
commit is further work from #7269

Signed-off-by: Brian L. Troutwine <brian@troutwine.us>

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -91,11 +91,22 @@ jobs:
...
             ~/.cargo/registry
             ~/.cargo/git
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-cargo-
+      - uses: actions/cache@v2.1.5
+        name: Cache Cargo build
+        with:
+          path: |
+            ~/target/debug
```
Comment (jszwedko) [inline .github/workflows/test.yml:106]: Is this right? I think you want just `target/debug`.

Comment (blt) [reply .github/workflows/test.yml:106]: Quite.

**Comment by blt** [comment]
Hmm, it looks like `krb5-src` does not build properly from cached artifacts.

**Comment by blt** [comment]
@jszwedko should be ready for another look when you have a chance. I figure I’ll need more time to figure out the debug cache issue and this is a marginal time improvement as is for repeat run PRs.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -91,11 +91,14 @@ jobs:
...
       - uses: actions/checkout@v2
       - uses: actions/cache@v2.1.5
+        name: Cache Cargo registry + index
         with:
           path: |
             ~/.cargo/registry
             ~/.cargo/git
           key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-cargo-
```
Comment (jszwedko) [inline .github/workflows/test.yml:101]: 🤔 what caches with this key?

Comment (blt) [reply .github/workflows/test.yml:101]: It's a prefix. So, previous runs that hash `Cargo.lock` will be reused and then updated in place. It saves a little time pulling mostly the same deps.

Comment (blt) [reply .github/workflows/test.yml:101]: TBH I had to fiddle with it for a bit to understand their docs on the subject.

Comment (jszwedko) [reply .github/workflows/test.yml:101]: Aha, I see. It is certainly not clear that this is a prefix from their docs or the field name. 👍

Comment (jszwedko) [reply .github/workflows/test.yml:101]: Maybe we could add a comment to these?

**Review by jszwedko — APPROVED** [review]
Nice, thanks!

**Pull request merged**